File: Record.cpp
//
// Created by Bryan Lim on 25/9/24.
//

#include "Record.h"

// Constructor
Record::Record(std::string GAME_DATE_EST,
               long long TEAM_ID_home,
               short PTS_home,
               float FG_PCT_home, float FT_PCT_home, float FG3_PCT_home,
               short AST_home, short REB_home,
               bool HOME_TEAM_WINS)
    : GAME_DATE_EST(GAME_DATE_EST), TEAM_ID_home(TEAM_ID_home), PTS_home(PTS_home),
      FG_PCT_home(FG_PCT_home), FT_PCT_home(FT_PCT_home), FG3_PCT_home(FG3_PCT_home),
      AST_home(AST_home), REB_home(REB_home),
      HOME_TEAM_WINS(HOME_TEAM_WINS){}

// Getter and Setter for GAME_DATE_EST
std::string Record::getGameDateEst() const {
    return GAME_DATE_EST;
}

void Record::setGameDateEst(const std::string& value) {
    GAME_DATE_EST = value;
}

// Getter and Setter for TEAM_ID_home
long long Record::getTeamIdHome() const {
    return TEAM_ID_home;
}

void Record::setTeamIdHome(long long value) {
    TEAM_ID_home = value;
}

// Getter and Setter for PTS_home
short Record::getPtsHome() const {
    return PTS_home;
}

void Record::setPtsHome(short value) {
    PTS_home = value;
}

// Getter and Setter for FG_PCT_home
float Record::getFgPctHome() const {
    return FG_PCT_home;
}

void Record::setFgPctHome(float value) {
    FG_PCT_home = value;
}

// Getter and Setter for FT_PCT_home
float Record::getFtPctHome() const {
    return FT_PCT_home;
}

void Record::setFtPctHome(float value) {
    FT_PCT_home = value;
}

// Getter and Setter for FG3_PCT_home
float Record::getFg3PctHome() const {
    return FG3_PCT_home;
}

void Record::setFg3PctHome(float value) {
    FG3_PCT_home = value;
}

// Getter and Setter for AST_home
short Record::getAstHome() const {
    return AST_home;
}

void Record::setAstHome(short value) {
    AST_home = value;
}

// Getter and Setter for REB_home
short Record::getRebHome() const {
    return REB_home;
}

void Record::setRebHome(short value) {
    REB_home = value;
}

// Getter and Setter for HOME_TEAM_WINS
bool Record::getHomeTeamWins() const {
    return HOME_TEAM_WINS;
}

void Record::setHomeTeamWins(bool value) {
    HOME_TEAM_WINS = value;
}

// Calculate the size of record in bytes
int Record::size(){
    return sizeof(GAME_DATE_EST) + sizeof(TEAM_ID_home) + sizeof(PTS_home) + sizeof(FG_PCT_home) + sizeof(FT_PCT_home) + sizeof(FG3_PCT_home) + sizeof(AST_home) + sizeof(REB_home) + sizeof(HOME_TEAM_WINS);
};

void Record::display() const {
    std::cout << "Game Date: " << GAME_DATE_EST << std::endl;
    std::cout << "Team ID (Home): " << TEAM_ID_home << std::endl;
    //std::cout << "Points (Home): " << PTS_home << std::endl;
    std::cout << "Field Goal % (Home): " << FG_PCT_home << std::endl;
    //std::cout << "Free Throw % (Home): " << FT_PCT_home << std::endl;
    //std::cout << "3-Point Field Goal % (Home): " << FG3_PCT_home << std::endl;
    //std::cout << "Assists (Home): " << AST_home << std::endl;
    //std::cout << "Rebounds (Home): " << REB_home << std::endl;
    //std::cout << "Home Team Wins: " << (HOME_TEAM_WINS ? "Yes" : "No") << std::endl;
}

File: Record.java
import java.nio.ByteBuffer;

public class Record { // 25 bytes for attributes + 7 bytes padding because java memory alignment in 8
                      // bytes
    private int game_date_est; // 4 bytes
    private int team_id_home; // 4 bytes
    private short pts_home; // 2 bytes
    private float fg_pct_home; // 4 bytes
    private float ft_pct_home; // 4 bytes
    private float fg3_pct_home; // 4 bytes
    private byte ast_home; // 1 byte
    private byte reb_home; // 1 byte
    private byte home_team_wins; // 1 byte
    private byte[] padding; // 7 bytes padding for alignment, ensuring record is 32 bytes

    public Record(int game_date_est, int team_id_home, short pts_home, float fg_pct_home, float ft_pct_home,
            float fg3_pct_home, byte ast_home, byte reb_home, byte home_team_wins) {
        this.game_date_est = game_date_est;
        this.team_id_home = team_id_home;
        this.pts_home = pts_home;
        this.fg_pct_home = fg_pct_home;
        this.ft_pct_home = ft_pct_home;
        this.fg3_pct_home = fg3_pct_home;
        this.ast_home = ast_home;
        this.reb_home = reb_home;
        this.home_team_wins = home_team_wins;
        this.padding = new byte[7]; // 7 bytes of padding

    }

    // Method to serialize the Record into a byte array
    public byte[] toByteArray() {
        ByteBuffer buffer = ByteBuffer.allocate(Block.RECORD_SIZE); // 32 bytes
        buffer.putInt(game_date_est);
        buffer.putInt(team_id_home);
        buffer.putShort(pts_home);
        buffer.putFloat(fg_pct_home);
        buffer.putFloat(ft_pct_home);
        buffer.putFloat(fg3_pct_home);
        buffer.put(ast_home);
        buffer.put(reb_home);
        buffer.put(home_team_wins);
        buffer.put(padding); // Add 7 bytes of padding

        return buffer.array();
    }

    // Constructor to deserialize a Record from a byte array
    public Record(byte[] recordData) {
        ByteBuffer buffer = ByteBuffer.wrap(recordData);
        this.game_date_est = buffer.getInt();
        this.team_id_home = buffer.getInt();
        this.pts_home = buffer.getShort();
        this.fg_pct_home = buffer.getFloat();
        this.ft_pct_home = buffer.getFloat();
        this.fg3_pct_home = buffer.getFloat();
        this.ast_home = buffer.get();
        this.reb_home = buffer.get();
        this.home_team_wins = buffer.get();
        this.padding = new byte[7]; // Skip 7 bytes of padding
        buffer.get(this.padding);
    }

    // Getter for game_date_est
    public int getGame_date_est() {
        return game_date_est;
    }

    // Getter for team_id_home
    public int getTeam_id_home() {
        return team_id_home;
    }

    // Getter for pts_home
    public short getPts_home() {
        return pts_home;
    }

    // Getter for fg_pct_home
    public float getFg_pct_home() {
        return fg_pct_home;
    }

    // Getter for ft_pct_home
    public float getFt_pct_home() {
        return ft_pct_home;
    }

    // Getter for fg3_pct_home
    public float getFg3_pct_home() {
        return fg3_pct_home;
    }

    // Getter for ast_home
    public byte getAst_home() {
        return ast_home;
    }

    // Getter for reb_home
    public byte getReb_home() {
        return reb_home;
    }

    // Getter for home_team_wins
    public byte getHome_team_wins() {
        return home_team_wins;
    }

    public String getUniqueId() {
        return Integer.valueOf(game_date_est).toString() + Integer.valueOf(team_id_home).toString();
    }
}
